src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -7905,56 +7920,258 @@ def apply(
                 "Snowpark pandas apply API doesn't yet support DataFrame or Series in 'args' or 'kwargs' of 'func'"
             )
 
-        # get input types of all data columns from the dataframe directly
-        input_types = self._modin_frame.get_snowflake_type(
-            self._modin_frame.data_column_snowflake_quoted_identifiers
-        )
+        if axis == 0:
+            frame = self._modin_frame
 
-        from snowflake.snowpark.modin.pandas.utils import try_convert_index_to_native
+            # To apply function to Dataframe with axis=0, we repurpose the groupby apply function by taking each
```

Comment:
BUT, if a df has 10 columns, then this approach is going to create 10 UDTFs and run them sequentially which is slow. We can define a single UDTF to handle the whole df in one go. Why not?

Reply:
Yeah, there are different approaches, here I'm reusing existing building block (groupby_apply) as much as possible so we can start enabling for customers, at which point we can see what kind of workloads and performance our customers are encountering and prioritize further improvements accordingly.

While an uber UDTF approach would reduce the number of round-trip queries, it's not clear to what extent the overall performance would be better.  For instance, partitioning by columns (as done here) should enable better scalability for larger data sets since they operate on column fraction of the total dataset at a time, and further these queries could be parallelized so as to substantially decreasing the overall latency.  While a single uber udtf approach would need the entire dataset in memory at one time but take similar overall execution time without the parallelization opportunity being in a single python method.

Reply:
No doubt this approach is ready to use at least.

The performance of a single UDTF approach should be better but I don't have the data to show how much it can be better IMO. Taking the 10 column case as an example, the current approach basically run 10 queries and each one UDTF basically runs the select query and apply the UDTF on the selected row sets. If we can combine them into one query and then the backend automatically will parallel the process for each rowsets across the warehouse instances.

But it's probably too early to optimize right now. It's better to add some perf tests for apply axis=0 first. cc @sfc-gh-nkumar @sfc-gh-helmeleegy @sfc-gh-lyambem

Reply:
I've created a jira https://snowflakecomputing.atlassian.net/browse/SNOW-1659582 to follow up on evaluating performance options when we have time.

Also had some side discussion with @sfc-gh-helmeleegy will share for visibility:

![Screenshot 2024-09-10 at 2 22 32 PM](https://github.com/user-attachments/assets/f3997f4e-abe0-4e5b-aa52-8cec9d9e9b3b)